File: scripts/force_sync_lyrics.py
import os
import sqlite3
import glob

# Paths based on the root directory
PROJECT_ROOT = r"d:\PersonalProject\MoodyMusic"
DB_PATH = os.path.join(PROJECT_ROOT, "storage", "db", "moody.db.bak") # Updating the local backup first to avoid messing up local dev db
LYRICS_DIR = os.path.join(PROJECT_ROOT, "storage", "lyrics")

print(f"Connecting to database: {DB_PATH}")
conn = sqlite3.connect(DB_PATH)
cursor = conn.cursor()

# Find all .lrc files in the lyrics directory
lrc_files = glob.glob(os.path.join(LYRICS_DIR, "**", "*.lrc"), recursive=True)
updated_count = 0

for file_path in lrc_files:
    # Example: d:\...\storage\lyrics\周杰伦\Jay\l_23291.lrc
    filename = os.path.basename(file_path)
    if filename.startswith("l_") and filename.endswith(".lrc"):
        # Extract ID
        id_str = filename[2:-4]
        try:
            song_id = int(id_str)
            # Calculate relative path
            rel_path = os.path.relpath(file_path, LYRICS_DIR)
            rel_path_unix = rel_path.replace("\\", "/")
            
            # Check if song exists
            cursor.execute("SELECT id, title, lrc_path FROM songs WHERE id = ?", (song_id,))
            song = cursor.fetchone()
            if song:
                current_lrc = song[2]
                if current_lrc != rel_path_unix:
                    print(f"Updating song [{song_id}] - {song[1]}: {current_lrc} -> {rel_path_unix}")
                    cursor.execute("UPDATE songs SET lrc_path = ? WHERE id = ?", (rel_path_unix, song_id))
                    updated_count += 1
        except ValueError:
            print(f"Skipping invalid filename: {filename}")

conn.commit()
conn.close()
print(f"Successfully updated {updated_count} lyrics mappings in the database.")

File: backend/cmd/tools/force_sync_lyrics.go
package main

import (
	"database/sql"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"strings"

	"moody-backend/internal/database"
)

func main() {
	// Paths based on the root directory
	projectRoot := `d:\PersonalProject\MoodyMusic`
	dbPath := filepath.Join(projectRoot, "storage", "db", "moody.db.bak")
	lyricsDir := filepath.Join(projectRoot, "storage", "lyrics")

	fmt.Printf("Connecting to database: %s\n", dbPath)
	database.InitDB(dbPath)
	defer database.DB.Close()

	updatedCount := 0

	err := filepath.Walk(lyricsDir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if !info.IsDir() && strings.HasPrefix(info.Name(), "l_") && strings.HasSuffix(info.Name(), ".lrc") {
			filename := info.Name()
			idStr := filename[2 : len(filename)-4]
			var songID int64
			if _, err := fmt.Sscanf(idStr, "%d", &songID); err == nil {
				relPath, _ := filepath.Rel(lyricsDir, path)
				relPathUnix := filepath.ToSlash(relPath)

				var title, currentLrc string
				var currentLrcNull sql.NullString
				err = database.DB.QueryRow("SELECT title, lrc_path FROM songs WHERE id = ?", songID).Scan(&title, &currentLrcNull)
				if err == nil {
					currentLrc = currentLrcNull.String
					if currentLrc != relPathUnix {
						fmt.Printf("Updating song [%d] - %s: %s -> %s\n", songID, title, currentLrc, relPathUnix)
						_, err = database.DB.Exec("UPDATE songs SET lrc_path = ? WHERE id = ?", relPathUnix, songID)
						if err != nil {
							fmt.Printf("Failed to update song [%d]: %v\n", songID, err)
						} else {
							updatedCount++
						}
					}
				}
			}
		}
		return nil
	})

	if err != nil {
		log.Fatalf("Error walking lyrics directory: %v", err)
	}

	fmt.Printf("Successfully updated %d lyrics mappings in the database.\n", updatedCount)
}
